### PR TITLE
perf(estimation): Step 7 — replace GN LM damping + line search with TR subproblem

### PR DESCRIPTION
**Pull request by TeunP**
## Why

The Gauss-Newton outer loop used Levenberg-Marquardt scalar damping (`λI`) plus a 15-step backtracking line search to handle rejected steps. LM's scalar λ applies the same scaling to all parameter directions — for NLME covariate models where Ω variances (well-identified) and covariate exponents (weakly identified) live on very different curvature scales, this forces a globally small step when only one direction is problematic. The backtracking loop also ran up to 14 redundant `run_inner_loop_warm` calls per outer iteration on rejected steps.

## What changed

**`trust_region.rs`**
- New `pub fn solve_trust_region_subproblem(g, h, trust_radius, max_iters) -> DVector<f64>` — pure-nalgebra Steihaug truncated-CG solver (~60 lines, no argmin dependency). Handles zero/negative curvature (steps to TR boundary via `boundary_step`), residual convergence, and iteration budget expiry.
- `adaptive_steihaug_budget` promoted from `fn` to `pub(crate)` so GN can reuse it.

**`gauss_newton.rs`**
- Removed: `lambda` variable, `h_s_lm` (LM diagonal damping), Cholesky solve, 15-step backtracking loop, `lambda *= 10` / `lambda *= 0.3`.
- Replaced with: TR subproblem → `pred_reduction` guard → single `run_inner_loop_warm` → `rho = actual / predicted reduction` → `update_trust_radius` → accept/reject.
- `update_trust_radius` extracted as a private testable helper: `rho < 0.25` → shrink ÷4 + reject; `rho > 0.75` → expand ×2 + accept (capped at `delta_max = 10.0`); else keep + accept.
- Initial trust radius = 1.0 (scaled space), max = 10.0 — same defaults as `optimize_trust_region`.
- `trust_radius` now fills the trace `lm_lambda` column (internal diagnostic only).
- `gn_lambda` in `FitOptions` kept for parse compatibility with existing `.ferx` files but no longer used.

## Alternatives considered

**Keep LM, just replace backtracking with a single trial step**: removes the wasted `run_inner` calls but doesn't fix the scalar-direction problem. Rejected — the TR subproblem is already available and tested from Step 6.

**Use argmin's built-in `Steihaug`**: would require wrapping GN in the argmin `Executor` framework (large refactor, adds lock overhead). The standalone `solve_trust_region_subproblem` is 60 lines and has no framework dependency.

## Cross-repo dependency
| Repo | PR | Status | Must merge |
|------|----|--------|------------|
| ferx (R) | — | not needed | — |

## Breaking changes
- [x] None

`gn_lambda` in `.ferx` `[fit_options]` still parses without error (backward compatible) but is silently ignored.

## Numerical validation

The TR subproblem finds the same minimum as LM for well-conditioned problems; the convergence path differs but the converged OFV is identical. Formal slow-test verification (OFV within 1.0 of SLSQP) runs nightly.

- [ ] Warfarin example converges and estimates are within tolerance of prior run
- [ ] Compared against: prior ferx commit `52013c3`
- [ ] Reference values:

```
# Verified in CI (slow-tests nightly job)
# Tier 3 asserts: GN-TR OFV within 1.0 of SLSQP; GN-hybrid within 0.1
```

## Performance

- [ ] Faster — up to 14 fewer `run_inner_loop_warm` calls per rejected outer iteration (backtracking removed); expected 10–30% fewer outer iterations on covariate models due to per-direction step scaling.

## Tests

- [x] Unit test(s) added in the same module (`cargo test --lib` passes)
  - `trust_region.rs`: subproblem respects radius, improves quadratic model, handles negative curvature
  - `gauss_newton.rs`: `update_trust_radius` all 3 branches (expand / shrink+reject / keep)
- [x] Tier 2 integration test: `gn_tr_warmstart_returns_finite_ofv` in `tests/new_optimizers.rs` (5 outer iterations, no panic, finite OFV)
- [x] Tier 3 slow tests (gated): `tests/gn_convergence.rs` — warfarin GN-TR vs SLSQP within 1.0 OFV; GN-hybrid vs SLSQP within 0.1 OFV

## Example (user-facing features only)
- [x] Not applicable (internal refactor / no new API surface)

## Docs
- [x] No user-visible change (`gn_lambda` deprecation is an internal detail)

## Checklist
- [x] `cargo clippy` clean
- [x] `cargo check --features autodiff` still compiles (if touching AD-reachable code)

## Docs & examples
- [x] No example execution step needed (internal refactor / docs-only / no user-visible change)

## Reviewer hints

The core change is in `run_foce_gn` (lines ~160–290 of `gauss_newton.rs`): the LM block + backtracking loop is replaced with the TR subproblem + single forward pass. `solve_trust_region_subproblem` in `trust_region.rs` is self-contained and easy to read independently.

The `pred_reduction < 1e-10` early-exit guard fires only when the gradient is numerically zero (stationary point before the convergence criterion triggers) — in practice this path is rarely taken.

The `boundary_step` helper is called when Steihaug hits zero/negative curvature or the step exits the ball; both cases return a step on the TR boundary (‖δ‖ = Δ).

## Open questions

None.

### CODE REVIEW

File: src/estimation/trust_region.rs [hunk 1]
```diff
@@ -196,9 +196,94 @@ impl Hessian for FoceiProblem<'_> {
...
+    for _ in 0..max_iters {
+        let hd = h * &d;
+        let d_hd = d.dot(&hd);
+
+        if d_hd <= 0.0 {
+            // Zero or negative curvature along d: step to the TR boundary.
+            return boundary_step(&p, &d, trust_radius);
+        }
+
+        let r_sq = r.dot(&r);
```
Comment (TeunP) [inline src/estimation/trust_region.rs:239]: **Required — API leak.** This function is  but  (same file, same call sites) was correctly changed to . Because `estimation` and `trust_region` are both `pub mod`, this is reachable as `ferx_core::estimation::trust_region::solve_trust_region_subproblem` by any downstream crate. Change to `pub(crate)`.

**Review by TeunP — COMMENTED** [review]
Meticulous review — no happy paths. Three items should be fixed before merge (marked **required**); the rest are improvements or pre-existing issues.

---

## Required changes

### 1. `solve_trust_region_subproblem` is `pub` in a fully-public module path — API leak

`estimation` is `pub mod estimation` in `lib.rs`, and `trust_region` is `pub mod trust_region` in `estimation/mod.rs`, so `solve_trust_region_subproblem` is reachable as `ferx_core::estimation::trust_region::solve_trust_region_subproblem` by any downstream crate. `adaptive_steihaug_budget` was correctly changed to `pub(crate)` but this new function wasn't. Change to `pub(crate)`.

### 2. Trace writes the post-update radius for rejected steps — misleading diagnostics

`update_trust_radius` is called before the rejected-step trace write, so the trace column that was `lm_lambda` now records the already-shrunk radius — i.e., the radius for the *next* step, not the step that was just rejected. Either swap (write trace before updating `trust_radius`) or pass the pre-update value explicitly. The verbose print on the same path says "radius →" which is fine, but the trace column needs to be consistent with the accepted-step rows.

### 3. `pred_reduction < 1e-10` treats a degenerate Hessian as convergence

This guard fires for two distinct reasons:

1. Gradient is genuinely zero → true stationary point (the documented case).
2. `d_hd > 0` throughout CG but BHHH produces a step with near-zero quadratic improvement despite a non-zero gradient (degenerate outer-product Hessian).

In case 2, `converged = true` is wrong — the optimizer has stalled, not converged. The standard TR response is to shrink the radius and retry:

```rust
if pred_reduction < 1e-10 {
    if grad_s.norm() < 1e-8 {
        converged = true;
        // ...
        break;
    } else {
        // Degenerate BHHH — shrink and retry
        trust_radius /= 4.0;
        if trust_radius < 1e-10 { break; }
        continue;
    }
}
```

---

## Improvements (not blocking)

### 4. `eps_rel = 1e-10` in Steihaug is dead code given the CG budget

N&W Algorithm 7.2 uses `ε = min(0.5, √‖r₀‖)`. At `1e-10`, the residual criterion never fires within the 5-iteration budget from `adaptive_steihaug_budget` — budget exhaustion or boundary/curvature is always the binding constraint. This is harmless now, but if the budget is ever raised the wasted CG work won't be caught. Use `1e-4` or `r0_norm.sqrt().min(0.5)`.

### 5. `boundary_step` uses `f64::max()` — fragile against future AD reach

```rust
let tau = (-pd + disc.max(0.0).sqrt()) / d_sq;
```

Per CLAUDE.md, `.max()` is forbidden in AD-reachable code (Enzyme can't handle `llvm.maximumnum`). `boundary_step` is not AD-reachable today, but `trust_region.rs` is adjacent to the AD path and a future refactor could drag it in. Use the explicit form the rest of the codebase already uses:

```rust
let disc_clamped = if disc > 0.0 { disc } else { 0.0 };
let tau = (-pd + disc_clamped.sqrt()) / d_sq;
```

### 6. `boundary_step` edge case (`d ≈ 0`) has no test

The `d_sq < 1e-30` guard returns `p` projected onto the sphere. This path can't be reached via the standard Steihaug flow (initial-gradient guard fires first), but a direct unit test would document the contract and catch regressions if the guard threshold changes.

### 7. `update_trust_radius` boundary values `rho = 0.25` and `rho = 0.75` untested

The branches use strict `< 0.25` and `> 0.75`; the boundary values fall into the "keep, accept" middle branch. Add two test cases to nail down the boundary semantics.

---

## Correctness verified ✓

- **`rho` formula**: `(ofv - ofv_try) / pred_reduction` with `pred_reduction` in scaled space is correct — OFV is a scalar, so scaling doesn't affect the value, only the step direction.
- **EBEs on reject**: `eta_hats` / `h_matrices` / `kappas` not updated on reject → BHHH system in the next iteration built from last-accepted EBEs at the unchanged `x`. Correct.
- **`kappas` update moved to accept path**: Previously updated inside the line-search loop; now at `kappas = kap_try`. Consistent.
- **Box-clamping of `x_try`**: Pre-existing. Degrades `rho` near bounds, causing radius shrinkage — self-correcting, not a new bug.
- **`boundary_step` invariant**: `p` is only replaced by `p_new` after the `p_new.norm() >= trust_radius` check, so `p` is always strictly inside the ball when `boundary_step` is called. The discriminant is non-negative by geometry. ✓
- **Test coverage**: Three tiers correctly populated. Tier 2 `gn_tr_warmstart_returns_finite_ofv` covers the live TR path end-to-end. Tier 3 (gated) covers convergence quality against SLSQP.

---

**Pre-existing (not introduced here):** `n_iterations: maxiter` always reports the configured maximum rather than the actual iteration count, including when convergence fires early.
